### PR TITLE
Migrate v3 programming model to v4 related with PR - #174

### PR DESCRIPTION
**Pull request by glaucia86**
Hi, folks. 

This issue is related with #174.

cc: @manekinekko @anfibiacreativa

### CODE REVIEW

**Comment by manekinekko** [comment]
@glaucia86  is this PR ready for a review?

**Comment by glaucia86** [comment]
> @glaucia86 is this PR ready for a review?

not yet, @manekinekko. Until the end of the day today. :)

**Comment by diberry** [comment]
@glaucia86 Do you think this needs API layer tests? Or do you see that as a separate chore?

File: packages/api-v4/src/index.ts
Comment (diberry) [inline packages/api-v4/src/index.ts:1]: Are you using anonymous because it works with azd and APIM?

File: packages/api-v4/src/models/favorite.schema.ts [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+ * author: Glaucia Lemos
+ * documentation reference: https://mongoosejs.com/docs/typescript.html
+ */
+
+import { model, Schema } from 'mongoose';
+
+export interface Favorite {
+  userId: string;
+  listingId: string;
+  createdAt?: string;
```
Comment (diberry) [inline packages/api-v4/src/models/favorite.schema.ts:14]: Did you want to add a default for this createdAt field?

File: packages/api-v4/src/models/listing.schema.ts [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+ * description: file responsible for the 'Listing' mongoose schema
+ * data: 07/27/2023
+ * author: Glaucia Lemos
+ * documentation reference: https://mongoosejs.com/docs/typescript.html
+ */
+
+import { ObjectId } from 'mongodb';
+import { model, Schema } from 'mongoose';
+
+export interface Listing {
```
Comment (diberry) [inline packages/api-v4/src/models/listing.schema.ts:12]: Could some of these have default values? review_stars, review_number, createdAt?

File: packages/api-v4/src/models/payment.schema.ts [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+ * description: file responsible for the 'Payment' mongoose schema
+ * data: 07/18/2023
+ * author: Glaucia Lemos
+ * documentation reference: https://mongoosejs.com/docs/typescript.html
+ */
+
+import { model, Schema } from 'mongoose';
+
+export interface Payment {
+  id: string;
```
Comment (diberry) [inline packages/api-v4/src/models/payment.schema.ts:12]: Could some of these have default values? provider, status, createdAt

File: packages/api-v4/src/models/payment.ts [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+ */
+
+import mongoose from "mongoose";
+import PaymentModel, { Payment } from "./payment.schema";
+
+export async function savePayment(payment: Partial<Payment>): Promise<Payment> {
+  return PaymentModel.create(payment);
+};
+
+export async function updatePaymentStatus(id: string, status: 'pending' | 'declined' | 'completed' | 'cancelled'): Promise<Payment | null> {
```
Comment (diberry) [inline packages/api-v4/src/models/payment.ts:15]: Can you pull status into its own type since its across several files?

File: packages/api-v4/src/models/reservation.schema.ts [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+ * file: packages/api-v4/src/models/reservation.schema.ts
+ * description: file responsible for the 'Reservation' mongoose schema
+ * data: 07/26/2023
+ * author: Glaucia Lemos
+ * documentation reference: https://mongoosejs.com/docs/typescript.html
+ */
+
+import { model, Schema } from 'mongoose';
+
+export interface Reservation {
```
Comment (diberry) [inline packages/api-v4/src/models/reservation.schema.ts:11]: Can any of these have default values?

**Review by diberry — APPROVED** [review]
All my comments are suggestions - ignore if they don't work with your goals for the project.

File: .vscode/tasks.json
[the remark this answers is not in the document]
Comment (glaucia86) [reply .vscode/tasks.json:1]: @manekinekko what do you think about it? Because in the first version looks like the `tasks.json` that was that way, right?

File: packages/api-v4/.env_example
[the remark this answers is not in the document]
Comment (glaucia86) [reply packages/api-v4/.env_example:1]: Yeah... good question. This `env_example` file I got from the `api` folder. As I saw in the code, we are using Azure Key Vault. But I have the same doubt as well: how the `local.settings.json` file can work together with `.env` file? 
Any thoughts about it @manekinekko?

File: packages/api-v4/host.json
[the remark this answers is not in the document]
Comment (glaucia86) [reply packages/api-v4/host.json:1]: Agree!

File: packages/api-v4/src/config/mongoose.ts [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+    if (database.readyState !== 1) {
+      console.log('Mongoose connecting...');
+      await mongoose.connect(config.connectionString, { dbName: config.database });
+      console.log('Mongoose connected successfully!');
+    }
+    else {
+      console.log('Mongoose already connected! Status:', database.readyState);
+    }
+  } catch (err) {
+    logger.error(`Mongoose database error: ${err}`);
```
[the remark this answers is not in the document]
Comment (glaucia86) [reply packages/api-v4/src/config/mongoose.ts:31]: return status of failure. There's a file called: `applicationInsightsTransports.ts` that I believe the intention of this file is to return all the status of failure from some certain service from the Application Insights. 
That was my read what i saw in the code

File: packages/api-v4/src/config/observability.ts [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+        client: applicationInsights.defaultClient,
+        level: LogLevel.Information,
+        handleExceptions: true,
+        handleRejections: true,
+      });
+
+      logger.add(applicationInsightsTransport);
+      logger.info('Added ApplicationInsights logger transport');
+  } catch (err) {
+    logger.error(`ApplicationInsights setup failed, ensure environment variable 'APPLICATIONINSIGHTS_CONNECTION_STRING' has been set. Error: ${err}`);
```
[the remark this answers is not in the document]
Comment (glaucia86) [reply packages/api-v4/src/config/observability.ts:66]: Same answer below

File: packages/api-v4/src/data/template/listings.json
[the remark this answers is not in the document]
Comment (glaucia86) [reply packages/api-v4/src/data/template/listings.json:1]: Good point

File: packages/api-v4/src/functions/favorites.ts [hunk 8]
```diff
@@ -0,0 +1,189 @@
...
+      listingId: listing.id,
+      userId: user.id,
+      createdAt: new Date().toISOString(),
+    });
+    return {
+      status: 200,
+      jsonBody: favoriteModel,
+    };
+  } catch (error) {
+    return {
```
[the remark this answers is not in the document]
Comment (glaucia86) [reply packages/api-v4/src/functions/favorites.ts:145]: Let me change here!

File: packages/api-v4/src/functions/favorites.ts [hunk 9]
```diff
@@ -0,0 +1,189 @@
...
+  }
+
+  try {
+    await removeFavorite({ listingId, userId });
+    return {
+      status: 204,
+    };
+  } catch (error: unknown) {
+    const err = error as Error;
+    console.error('Error...', err.message);
```
[the remark this answers is not in the document]
Comment (glaucia86) [reply packages/api-v4/src/functions/favorites.ts:182]: But with this new v4 can we still using `context`?

File: packages/api-v4/src/functions/httpTriggerTest.ts
[the remark this answers is not in the document]
Comment (glaucia86) [reply packages/api-v4/src/functions/httpTriggerTest.ts:1]: removed

File: packages/api-v4/src/functions/favorites.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (diberry) [reply packages/api-v4/src/functions/favorites.ts:182]: Yeah, you can see it in the default functions they create all have context.log at the top.

File: packages/api-v4/src/functions/payments.ts [hunk 10]
```diff
@@ -0,0 +1,158 @@
...
+import { findUserById } from '../models/user';
+import { updateReservationStatus } from '../models/reservation';
+import { Payment } from '../models/payment.schema';
+
+// GET: Get Payment by Id
+export async function getPaymentById(request: HttpRequest, context: InvocationContext): Promise<HttpResponseInit> {
+  context.log(`Http function getPaymentById processed request for url "${request.url}"`);
+
+  // uncomment if you decide to test the api in production (using database)
+  await initializeDatabaseConfiguration();
```
[the remark this answers is not in the document]
Comment (glaucia86) [reply packages/api-v4/src/functions/payments.ts:13]: You mean solve this way:

```ts
try {
    await initializeDatabaseConfiguration();
  } catch (error: unknown) {
    const err = error as Error;
    context.error(`Error initializing database configuration: ${err.message}`);
    return {
      status: 500,
      jsonBody: {
        error: 'Error initializing database configuration',
      },
    };
  }
  ```

File: packages/api-v4/src/functions/payments.ts [hunk 11]
```diff
@@ -0,0 +1,158 @@
...
+  const id = request.params.id ?? '';
+
+  const payment = await findPaymentById(id);
+
+  if (payment) {
+    return {
+      jsonBody: payment,
+    };
+  } else {
+    return {
```
[the remark this answers is not in the document]
Comment (glaucia86) [reply packages/api-v4/src/functions/payments.ts:24]: Or better: create a new folder called: `errros` with an `index.ts` file. And then create an error message for each one and then use accross the project. Same as **[here](https://github.com/rmanguinho/clean-ts-api/tree/master/src/presentation/errors)**

File: packages/api-v4/src/functions/payments.ts [hunk 12]
```diff
@@ -0,0 +1,158 @@
...
+      },
+    };
+  }
+};
+
+// POST: Create Payment
+export async function postPayment(request: HttpRequest, context: InvocationContext): Promise<HttpResponseInit> {
+  context.log(`Http function createPayment processed request for url "${request.url}"`);
+
+  await initializeDatabaseConfiguration();
```
[the remark this answers is not in the document]
Comment (glaucia86) [reply packages/api-v4/src/functions/payments.ts:102]: Same answer below: what do you think about that?

```typescript
try {
    await initializeDatabaseConfiguration();
  } catch (error: unknown) {
    const err = error as Error;
    context.error(`Error initializing database configuration: ${err.message}`);
    return {
      status: 500,
      jsonBody: {
        error: 'Error initializing database configuration',
      },
    };
  }
  ```

File: packages/api-v4/src/functions/reservations.ts [hunk 13]
```diff
@@ -0,0 +1,145 @@
...
+      }
+    }
+  }
+};
+
+// GET: Get Reservations
+export async function getReservations(request: HttpRequest, context: InvocationContext): Promise<HttpResponseInit> {
+  context.log(`Http function getReservations processed request for url "${request.url}"`);
+
+  const offset = Number(request.query.get('offset')) || 0;
```
[the remark this answers is not in the document]
Comment (glaucia86) [reply packages/api-v4/src/functions/reservations.ts:32]: All of them? Or just one's for offset and limit?

File: packages/api-v4/src/index.ts
Comment (glaucia86) [reply packages/api-v4/src/index.ts:1]: @manekinekko defined in the previous api as `anonymous`

File: packages/api-v4/src/functions/favorites.ts [hunk 14]
```diff
@@ -0,0 +1,189 @@
...
+};
+
+// DELETE: Favorite
+export async function deleteFavorite(request: HttpRequest, context: InvocationContext): Promise<HttpResponseInit> {
+  context.log(`Http function deleteFavorite processed request for url "${request.url}"`);
+
+  await initializeDatabaseConfiguration();
+
+  const jsonData = await request.json();
+  const { listingId, userId } = jsonData as { listingId: Favorite['listingId'], userId: Favorite['userId'] };
```
[the remark this answers is not in the document]
Comment (glaucia86) [reply packages/api-v4/src/functions/favorites.ts:164]: You mean this way?

```typescript
type DeleteFavoriteRequestBody = {
  listingId: Favorite['listingId'];
  userId: Favorite['userId'];
};

(... later in the deleteFavorite function)

  const jsonData = await request.json() as DeleteFavoriteRequestBody;

  const { listingId, userId } = jsonData;
```

File: packages/api-v4/src/functions/openapi.ts [hunk 15]
```diff
@@ -0,0 +1,74 @@
+import { HttpRequest, HttpResponseInit, InvocationContext } from "@azure/functions";
+import { existsSync, readFileSync } from "fs";
+import yaml from 'yamljs';
+import * as swaggerUi from 'swagger-ui-dist'
+
+export async function openApi(request: HttpRequest, context: InvocationContext): Promise<HttpResponseInit> {
```
[the remark this answers is not in the document]
Comment (glaucia86) [reply packages/api-v4/src/functions/openapi.ts:6]: good catch, @diberry

[the remark this answers is not in the document]
Comment (glaucia86) [reply packages/api-v4/src/functions/openapi.ts:6]: Done!

File: packages/api-v4/src/functions/payments.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (diberry) [reply packages/api-v4/src/functions/payments.ts:13]: Yes - you might want to abstract that out to its own function then use it across the app. Would this be more interesting as a middleware rejection?

File: packages/api-v4/src/functions/reservations.ts (hunk 13 above)
[the remark this answers is not in the document]
Comment (diberry) [reply packages/api-v4/src/functions/reservations.ts:32]: My understanding is that ?? is a tighter check than ||. But if that isn't correct, leave it as it is.

**Review by manekinekko — CHANGES_REQUESTED** [review]

File: packages/api-v4/.env_example [hunk 16]
```diff
@@ -0,0 +1,18 @@
...
+STRAPI_DATABASE_NAME=
+STRAPI_DATABASE_SSL=true
+
+# Mongo DB config
+MONGO_DATABASE_NAME=
+MONGO_CONNECTION_STRING=
+
+# App insight config
+APPINSIGHTS_INSTRUMENTATIONKEY=
+APPLICATIONINSIGHTS_CONNECTION_STRING=
```
[the remark this answers is not in the document]
Comment (glaucia86) [reply packages/api-v4/.env_example]: renamed :)

File: packages/api-v4/.gitignore [hunk 17]
```diff
@@ -0,0 +1,99 @@
...
+# Azure Functions artifacts
+bin
+obj
+appsettings.json
+local.settings.json
+
+# Azurite artifacts
+__blobstorage__
+__queuestorage__
+__azurite_db*__.json
```
[the remark this answers is not in the document]
Comment (glaucia86) [reply packages/api-v4/.gitignore:99]: changed

File: packages/api-v4/package.json [hunk 18]
```diff
@@ -0,0 +1,33 @@
...
+  "scripts": {
+    "build": "tsc",
+    "watch": "tsc -w",
+    "clean": "rimraf dist",
+    "prestart": "npm run clean && npm run build",
+    "start": "func start --typescript",
+    "test": "echo \"No tests yet...\""
+  },
+  "dependencies": {
+    "@azure/functions": "^4.0.0-alpha.7",
```
[the remark this answers is not in the document]
Comment (glaucia86) [reply packages/api-v4/package.json:18]: Updated. And @manekinekko I noticed that we have another packages outdated...

image

What do you think? Should update the minor one's and let the major's to update later?

File: packages/api-v4/src/config/appConfig.ts [hunk 19]
```diff
@@ -0,0 +1,28 @@
+/**
+ * file: packages/api-v4/src/config/appConfig.ts
+ * description: file responsible for the
+ * data: 07/05/2023
+ * author: Glaucia Lemos
+ */
+
```
[the remark this answers is not in the document]
Comment (glaucia86) [reply packages/api-v4/src/config/appConfig.ts]: Already removed

File: packages/api-v4/src/config/index.ts [hunk 20]
```diff
@@ -0,0 +1,88 @@
...
+    const credential = new DefaultAzureCredential();
+    const secretClient = new SecretClient(keyVaultEndpoint, credential);
+
+    for await (const secretProperties of secretClient.listPropertiesOfSecrets()) {
+      const secret = await secretClient.getSecret(secretProperties.name);
+
+      // KeyVault does not support underscores in key names and replaces '-' with '_'
+      // Expect KeyVault secret names to be in conventional capitalized snake casing after conversion
+      const keyName = secret.name.replace(/-/g, "_").toUpperCase();
+      process.env[keyName] = secret.value;
```
[the remark this answers is not in the document]
Comment (glaucia86) [reply packages/api-v4/src/config/index.ts:68]: Sorry, @manekinekko I didn't understand what you said here. Because in the oldest `api` folder you included the variable name as: `keyName` (https://github.com/Azure-Samples/contoso-real-estate/blob/main/packages/api/config/index.ts#L70)

So, should I include as sames as you defined here (`keyVaultName`): https://github.com/Azure-Samples/contoso-real-estate/blob/main/infra/app/api.bicep#L9C7-L9C20

[the remark this answers is not in the document]
Comment (glaucia86) [reply packages/api-v4/src/config/index.ts]: No problem ;)

File: packages/api-v4/src/functions/checkout.ts [hunk 21]
```diff
@@ -0,0 +1,195 @@
...
+export async function postCheckout(request: HttpRequest, context: InvocationContext): Promise<HttpResponseInit> {
+  context.log(`Http function postCheckout processed request for url "${request.url}"`);
+
+  await initializeDatabaseConfiguration();
+
+  const config = await getConfig();
+  const reservation = await request.json() as ReservationRequest;
+  const guests = Number(reservation.guests) || 0;
+
+  const from = new Date(reservation.from);
```
[the remark this answers is not in the document]
Comment (manekinekko) [reply packages/api-v4/src/functions/checkout.ts:18]: @glaucia86 let's create an issue to track this type of issues?

**Comment by manekinekko** [comment]
After investigation, it looks like `func` is not picking up the runtime, and hence not registering all functions. Here is the warning that's being triggered:

`WARNING: Failed to detect the Azure Functions runtime. Switching "@azure/functions" package to test mode - not all features are supported.`

image

**Comment by manekinekko** [comment]
Even after reading [this issue](https://github.com/microsoft/ApplicationInsights-node.js/issues/1144), and removing the `applicationinsights` package, the issue still persists.

I will keep on investigating.

**Comment by glaucia86** [comment]
> Even after reading [this issue](https://github.com/microsoft/ApplicationInsights-node.js/issues/1144), and removing the `applicationinsights` package, the issue still persists.
> 
> I will keep on investigating.

@manekinekko let me know when you have a time and then I schedule a time to see with you what is happening together

**Comment by ejizba** [comment]
You don't need to remove `applicationinsights`, you just need to update it.

Try running `npm update`. If that doesn't work, please post your package.json and package-lock.json here.

**Comment by glaucia86** [comment]
> You don't need to remove `applicationinsights`, you just need to update it.
> 
> Try running `npm update`. If that doesn't work, please post your package.json and package-lock.json here.

Hi, @ejizba! I ran the command: npm update and the problem persists. Below is the requested information:

<details><summary><b>packages/api-v4/package.json</b></summary>
<br/>

```json

{
  "name": "api-v4",
  "version": "1.0.0",
  "description": "A v4 programming model Azure Functions application.",
  "scripts": {
    "env": "azd env get-values --no-prompt > .env",
    "build": "tsc",
    "watch": "tsc -w",
    "clean": "rimraf dist",
    "prestart": "npm run clean && npm run build",
    "start:host": "func start --typescript",
    "start": "npm-run-all --parallel start:host watch",
    "lint": "eslint --config ../../.eslintrc.js",
    "lint:fix": "npm run lint -- --fix",
    "test": "echo \"No tests yet...\""
  },
  "dependencies": {
    "@azure/functions": "^4.0.0-alpha.11",
    "@azure/identity": "^3.2.3",
    "@azure/keyvault-secrets": "^4.7.0",
    "applicationinsights": "^2.7.0",
    "dotenv": "^16.3.1",
    "mongodb": "^5.6.0",
    "mongoose": "^7.3.1",
    "swagger-ui-dist": "^4.14.0",
    "yamljs": "^0.3.0",
    "pg": "^8.11.1",
    "winston": "^3.9.0",
    "winston-transport": "^4.5.0"
  },
  "devDependencies": {
    "@azure/functions": "^1.2.3",
    "@types/node": "^18.16.19",
    "@types/pg": "^8.10.2",
    "@types/swagger-ui-dist": "^3.30.1",
    "@types/yamljs": "^0.2.31",
    "eslint": "^8.31.0",
    "eslint-config-prettier": "^8.6.0",
    "@typescript-eslint/eslint-plugin": "^5.48.1",
    "@typescript-eslint/parser": "^5.48.1",
    "npm-run-all": "^4.1.5",
    "azure-functions-core-tools": "^4.0.5198",
    "rimraf": "^5.0.0",
    "typescript": "^4.8.4"
  },
  "main": "dist/src/index.js"
}
```
</details>
<br/>

<details><summary><b>package.json (from the root of the project)</b></summary>
<br/>

```json
{
  "name": "contoso-real-estate",
  "version": "1.0.0",
  "description": "",
  "private": true,
  "scripts": {
    "start": "concurrently npm:start:* --kill-others",
    "start:services": "docker compose up",
    "start:api": "npm run start --workspace=api-v4",
    "start:website": "npm run start:swa --workspace=portal",
    "test": "npm run test -ws --if-present",
    "build": "npm run build -ws --if-present",
    "format": "prettier --write .",
    "format:check": "prettier --check .",
    "lint": "npm run lint -ws --if-present",
    "lint:fix": "npm run lint:fix -ws --if-present",
    "clean": "rimraf \"packages/**/*.tsbuildinfo\"",
    "clean:install": "rimraf \"packages/**/node_modules\" \"node_modules\" && npm install"
  },
  "keywords": [],
  "author": "Microsoft",
  "license": "MIT",
  "workspaces": [
    "packages/*"
  ],
  "engines": {
    "node": "18.15.0",
    "npm": ">=8"
  },
  "devDependencies": {
    "concurrently": "^7.6.0",
    "git-commit-msg-linter": "^4.2.1",
    "prettier": "^2.7.1",
    "rimraf": "^4.3.1"
  },
  "prettier": {
    "trailingComma": "all",
    "tabWidth": 2,
    "semi": true,
    "singleQuote": false,
    "printWidth": 120,
    "bracketSpacing": true,
    "arrowParens": "avoid"
  }
}
```
</details>
<br/>

And this is the `package-lock.json` file from the root of the project (gist)
**[HERE](https://gist.github.com/glaucia86/2b4b597eaf6e4c39d6b30a178c59e261)**

**Comment by glaucia86** [comment]
@manekinekko and @ejizba 

The API is now returning (as shown in the images below)

Captura de Tela 2023-09-08 às 19 28 41

Captura de Tela 2023-09-08 às 19 36 18

What did I have to do? Follow Eric Jizba's guidance to update the applicationinsights package and also include the `local.settings.json` configuration in the root of the `api-v4` folder:

* local.settings.json

```json
{
   "IsEncrypted": false,
   "Values": {
     "FUNCTIONS_WORKER_RUNTIME": "node",
     "AzureWebJobsFeatureFlags": "EnableWorkerIndexing"
   }
}
```

**Comment by manekinekko** [comment]
Hi @glaucia86 can you give me write access to [your repo](https://github.com/glaucia86/contoso-real-estate/tree/feat/migrate-v3-to-v4-functions), I need to send a few changes. Thx

**Comment by glaucia86** [comment]
> manekinekko

Sent you now! Thanks a lot, Wassim! :)

**Comment by manekinekko** [comment]
@glaucia86 can you apply these changes? I cannot push to your rep. Thanks.

<details>
<summary>diff.patch</summary>

```diff
diff --git a/packages/api-v4/src/functions/favorites.ts b/packages/api-v4/src/functions/favorites.ts
index f8b96b2..b992e99 100644
--- a/packages/api-v4/src/functions/favorites.ts
+++ b/packages/api-v4/src/functions/favorites.ts
@@ -3,12 +3,7 @@ import { initializeDatabaseConfiguration } from "../config";
 import { fetchFavoritesDataByUserId, findFavorite, getFavoritesByUserId, saveFavorite, removeFavorite } from '../models/favorite';
 import { User } from "../models/user.schema";
 import { Listing } from "../models/listing.schema";
-import { Favorite } from "../models/favorite.schema";
 
-type DeleteFavoriteRequestBody = {
-  listingId: Favorite['listingId'];
-  userId: Favorite['userId'];
-};
 
 // GET: Favorites
 export async function getFavorites(request: HttpRequest, context: InvocationContext): Promise<HttpResponseInit> {
@@ -16,8 +11,9 @@ export async function getFavorites(request: HttpRequest, context: InvocationCont
 
   const offset = Number(request.query.get('offset')) || 0;
   const limit = Number(request.query.get('limit')) || 10;
-
-  const { userId, listingId, aggregate } = request.params;
+  const userId = request.query.get('userId');
+  const listingId = request.query.get('listingId');
+  const aggregate = request.query.get('aggregate');
 
   // UserID is the only required parameter
   if (!userId || userId === 'undefined') {
@@ -166,8 +162,8 @@ export async function deleteFavorite(request: HttpRequest, context: InvocationCo
 
   await initializeDatabaseConfiguration();
 
-  const jsonData = await request.json() as DeleteFavoriteRequestBody;
-  const { listingId, userId } = jsonData;
+  const userId = request.query.get('userId');
+  const listingId = request.query.get('listingId');
 
   if (!listingId || !userId) {
     return {
diff --git a/packages/portal/package.json b/packages/portal/package.json
index e390817..d0f84a1 100644
--- a/packages/portal/package.json
+++ b/packages/portal/package.json
@@ -24,6 +24,7 @@
     "@angular/platform-browser": "^16.0.5",
     "@angular/platform-browser-dynamic": "^16.0.5",
     "@angular/router": "^16.0.5",
+    "@apollo/client": "^3.8.3",
     "apollo-angular": "^4.2.1",
     "graphql": "^16.6.0",
     "rxjs": "~7.5.0",

```

</details>

**Comment by glaucia86** [comment]
> @glaucia86 can you apply these changes? I cannot push to your rep. Thanks.
> 
> diff.patch

@manekinekko, finally done! 😀

**Comment by manekinekko** [comment]
@glaucia86 did you commit these changes to this PR? I don't see the new commit 😅

**Comment by glaucia86** [comment]
> @glaucia86 did you commit these changes to this PR? I don't see the new commit 😅

@manekinekko now is finally done!

File: infra/core/database/postgresql/flexibleserver.bicep [hunk 22]
```diff
@@ -75,5 +75,4 @@ resource keyVault 'Microsoft.KeyVault/vaults@2022-07-01' existing = {
   name: keyVaultName
 }
 
-output POSTGRES_SERVER_NAME string = postgresServer.name
```
Comment (manekinekko) [inline infra/core/database/postgresql/flexibleserver.bicep:78]: why these changes in this file?

File: infra/main.bicep [hunk 23]
```diff
@@ -404,7 +404,5 @@ output STRAPI_DATABASE_USERNAME string = cmsDatabaseUser
 output STRAPI_DATABASE_HOST string = cmsDB.outputs.POSTGRES_DOMAIN_NAME
 output STRAPI_DATABASE_PORT string = cmsDatabasePort
 
-output CMS_DATABASE_SERVER_NAME string = cmsDB.outputs.POSTGRES_SERVER_NAME
```
Comment (manekinekko) [inline infra/main.bicep:407]: why these changes in this file?

File: packages/portal/package.json [hunk 24]
```diff
@@ -24,11 +24,14 @@
...
     "@angular/platform-browser-dynamic": "^16.0.5",
     "@angular/router": "^16.0.5",
+    "@apollo/client": "^3.8.3",
+    "apollo-angular": "^4.2.1",
     "@fortawesome/angular-fontawesome": "^0.13.0",
     "@fortawesome/fontawesome-svg-core": "^6.4.2",
     "@fortawesome/free-brands-svg-icons": "^6.4.2",
     "@fortawesome/free-solid-svg-icons": "^6.4.2",
     "apollo-angular":  "^5.0.1",
+
```
Comment (manekinekko) [inline packages/portal/package.json:34]:
```suggestion
```

File: scripts/database/restore.sh [hunk 25]
```diff
@@ -30,27 +30,13 @@ else
...
-  az postgres flexible-server firewall-rule create \
-    --resource-group "rg-$AZURE_ENV_NAME" \
-    --name "$CMS_DATABASE_SERVER_NAME" \
-    --rule-name "AllowMyIP" \
-    --start-ip-address "$my_public_ip" \
-    --end-ip-address "$my_public_ip" \
-    --output none
-fi
-
 filename="${1:-}"
```
Comment (manekinekko) [inline scripts/database/restore.sh:40]: why these changes in this file?

**Review by manekinekko — CHANGES_REQUESTED** [review]

**Comment by manekinekko** [comment]
@glaucia86 can you also add a readme.md file under packages/api-v4/ explaning how to start the new API? (similarily to packages/api/README.md). Some steps are necessary for the new API to execute, for instance renaming `local.settings.sample.json` to `local.settings.json`, etc...

File: infra/core/database/postgresql/flexibleserver.bicep (hunk 22 above)
Comment (glaucia86) [reply infra/core/database/postgresql/flexibleserver.bicep:78]: Uhhh.... I just did git rebase from the main.... let me remove

File: scripts/database/restore.sh (hunk 25 above)
Comment (glaucia86) [reply scripts/database/restore.sh:40]: Uhhh.... I just did git rebase from the main.... let me remove

File: packages/api-v4/README.md [hunk 26]
```diff
@@ -0,0 +1,344 @@
+# Instructions for using the API-V4
```
Comment (manekinekko) [inline packages/api-v4/README.md:1]: can you add this text to the beginning of the readme?

<pre>
# Contoso Real Estate App: API Package (v4)

**IMPORTANT: THIS REPOSITORY IS OPTIMIZED FOR CODESPACES AND TO WORK AS A SET OF COMPOSABLE APPS AND APIS. STANDALONE PACKAGE FUNCTIONALITY IS LIMITED AND MAY REQUIRE ADDITIONAL CONFIGURATION OR DEVELOPMENT**

This package deploys an Azure Functions API that is used by the Contoso Real Estate App, with multiple endpoints, and serving multiple applications, that are part of the scenarios.

## Pre-requisites

**IMPORTANT: THIS SCENARIO IS TIGHTLY COUPLED WITH SCENARIO 3. SOME PARTS OF THIS FUNCTION APP MAY NOT WORK AS EXPECTED IF YOU DON'T FOLLOW THE INSTRUCTIONS IN SCENARIO 3. THIS FUNCTION APP QUERIES TWO DATABASES. THEY MAY NEED TO BE IN PLACE IN ORDER FOR THE ENDPOINTS TO WORK**

If you want to run the API independently and locally, the following technologies must be in place:

- [node.js](https://nodejs.org) LTS, with the corresponding npm version
- [Azure Core Tools](https://learn.microsoft.com/azure/azure-functions/functions-run-local)

</pre>

File: packages/api-v4/README.md [hunk 27]
```diff
@@ -0,0 +1,344 @@
...
+      "updated_by_id":null
+   }
+]
+```
+</details>
+<br/>
+
+If you have any questions, don't hesitate to contact me by opening an issue directly in the repository.
+
+
```
Comment (manekinekko) [inline packages/api-v4/README.md:344]: This is not necessary.

File: packages/api-v4/README.md [hunk 28]
```diff
@@ -0,0 +1,344 @@
...
+```json
+{
+  "IsEncrypted": false,
+  "Values": {
+    "FUNCTIONS_WORKER_RUNTIME": "node",
+    "AzureWebJobsFeatureFlags": "EnableWorkerIndexing"
+  }
+}
+```
+
```
Comment (manekinekko) [inline packages/api-v4/README.md:27]: Can you include the same intruction as for the `packages/api/`?

<pre>

## Steps to start the API

1. fork or clone the repository locally
2. assuming you are in the folder containing your clone, go to the terminal and run

```bash
cd packages/api-v4
```

3. create the `local.settings.json` file and add the following block of code:

```json
{
  "IsEncrypted": false,
  "Values": {
    "FUNCTIONS_WORKER_RUNTIME": "node",
    "AzureWebJobsFeatureFlags": "EnableWorkerIndexing"
  }
}
```

4. Now install all the dependencies for the API and run it

```bash
npm install && npm start
```
</pre>

File: packages/api-v4/README.md [hunk 29]
```diff
@@ -0,0 +1,344 @@
...
+* Flexible folder structure
+* Being able to define function.json directly in the function's in the code
+* New HTTP trigger types
+* Improved IntelliSense
+* Timer Trigger (TypeScript)
+* Durable Functions (TypeScript)
+  
+I wrote an article explaining how this whole migration process went and how you can do the same. You can read the article [here](https://techcommunity.microsoft.com/t5/educator-developer-blog/step-by-step-guide-migrating-v3-to-v4-programming-model-for/ba-p/3897691).
+
```
Comment (manekinekko) [inline packages/api-v4/README.md:13]: This is not necessary.

File: infra/core/database/postgresql/flexibleserver.bicep [hunk 30]
```diff
@@ -76,4 +76,3 @@ resource keyVault 'Microsoft.KeyVault/vaults@2022-07-01' existing = {
 }
 
 output POSTGRES_SERVER_NAME string = postgresServer.name
-output POSTGRES_DOMAIN_NAME string = postgresServer.properties.fullyQualifiedDomainName
```
Comment (manekinekko) [inline infra/core/database/postgresql/flexibleserver.bicep:79]: It looks like this line got also deleted. Can you revert it?

**Review by manekinekko — CHANGES_REQUESTED** [review]

**Comment by glaucia86** [comment]
@manekinekko I made the changes as asked. Thanks a lot :)

**Comment by gabrieljoelc** [comment]
great work and blog post!

I wonder if it would have been easier to understand the v3 to v4 differences if the formatting changes were included in a separate PR

File: packages/api-v4/src/index.ts [hunk 31]
```diff
@@ -0,0 +1,118 @@
...
+import { getUsers, getUserById, postUsers } from "./functions/users";
+import { getPaymentById, getPayments, postPayment } from "./functions/payments";
+import { getReservationById, getReservations, patchReservationById } from "./functions/reservations";
+import { getFavorites, postFavorites, deleteFavorite } from "./functions/favorites";
+import { getListingById, getListings } from "./functions/listings";
+import { openApi } from "./functions/openapi";
+import { postCheckout } from "./functions/checkout";
+
+//#region User Function
+app.get("get-users", {
```
Comment (gabrieljoelc) [inline packages/api-v4/src/index.ts:11]: I love this `index.ts` file aggregating the bindings from the functions. It loosens the coupling between the endpoints and the Azure Function infrastructure a bit.

It also makes it easier to see the possibility of generating functions or bindings based on convention. I've always wished we could Railsify the creation of Azure Functions.

**Comment by glaucia86** [comment]
> * Question: how did you generate the `openapi.yml` file?
> * Feedback: I wonder if it would have been easier to understand the v3 to v4 differences if the formatting changes were included in a separate PR

Hi, @gabrieljoelc. How are you?

Unfortunately, it was a feature that should have been delivered as soon as possible due to the recent release. However, if you'd like to know more details about how a particular function was implemented, you can check out this blog I wrote with a step-by-step guide: **[HERE](https://techcommunity.microsoft.com/t5/educator-developer-blog/step-by-step-guide-migrating-v3-to-v4-programming-model-for/ba-p/3897691)**

About the `openapi.yaml` file, it was created by @manekinekko, and he would be better equipped to provide a response than I am.

**Review by anfibiacreativa — APPROVED** [review]
